build-logic/src/main/kotlin/dokkabuild/tasks/GitCheckoutTask.kt

```diff
@@ -0,0 +1,116 @@
+import org.eclipse.jgit.lib.ProgressMonitor
+import org.eclipse.jgit.lib.RepositoryCache
+import org.eclipse.jgit.lib.TextProgressMonitor
+import org.eclipse.jgit.util.FS
+import org.gradle.api.DefaultTask
+import org.gradle.api.file.DirectoryProperty
+import org.gradle.api.file.FileSystemOperations
+import org.gradle.api.provider.Property
+import org.gradle.api.tasks.CacheableTask
+import org.gradle.api.tasks.Input
+import org.gradle.api.tasks.OutputDirectory
+import org.gradle.api.tasks.TaskAction
+import java.io.File
+import javax.inject.Inject
+
+/**
+ * Clones a remote Git repository into [GitCheckoutTask.destination].
+ */
+@CacheableTask
+abstract class GitCheckoutTask @Inject constructor(
+    private val fs: FileSystemOperations
+) : DefaultTask() {
+
+    @get:OutputDirectory
+    abstract val destination: DirectoryProperty
+
+    /** Public HTTP URI of the Git repo. */
+    @get:Input
+    abstract val uri: Property<String>
+
+    /** Specific git hash to checkout. */
+    @get:Input
+    abstract val commitId: Property<String>
+
+    private val localRepoDir: File
+        get() = temporaryDir.resolve("repo")
+
+    private val gitOperationsPrinter: ProgressMonitor =
+        if (logger.isInfoEnabled) {
+            TextProgressMonitor()
+        } else {
+            NullProgressMonitor.INSTANCE
+        }
+
+    init {
+        group = "git checkout"
+    }
+
+    @TaskAction
+    fun action() {
+        initializeRepo()
+
+        fs.sync {
+            from(localRepoDir)
+            into(destination)
+            // exclude the .git dir:
+            // - prevent the root git repo getting confused with a nested git repo
+            // - improves Gradle caching
+            exclude(".git/")
+        }
+
+        logger.lifecycle("initialized git repo $uri in ${destination.asFile.orNull}")
+    }
+
+    /**
+     * Initialized [uri] in [localRepoDir], forcibly resetting it and deleting any untracked files or changes.
+     */
+    private fun initializeRepo() {
+        val uri = uri.get()
+        val commitId = commitId.get()
+
+        val gitRepoInitialized = RepositoryCache.FileKey.isGitRepository(localRepoDir, FS.DETECTED)
+
+        val repo = if (gitRepoInitialized) {
+            Git.open(localRepoDir)
+        } else {
+            fs.delete { delete(localRepoDir) }
+
+            Git.cloneRepository()
+                .setProgressMonitor(gitOperationsPrinter)
+                .setNoCheckout(true)
+                .setURI(uri)
+                .setDirectory(localRepoDir)
+                .call()
+        }
+
+        repo.use { git ->
+            git.checkout()
+                .setProgressMonitor(gitOperationsPrinter)
+                .setForced(true)
+                .setName(commitId)
+                .call()
+
+            // git reset --hard
+            git.reset()
+                .setProgressMonitor(gitOperationsPrinter)
+                .setMode(HARD)
+                .call()
+
+            // git clean -fdx
```

Comment:
Probably I should add the same for Kotlin build regression benchmarks :+1:

Reply:
which bit? :) My changes followed your example

https://github.com/JetBrains/kotlin/blob/f262247b2bcd7cc0e9d64d274f2fd412031187ab/libraries/tools/gradle/regression-benchmark-templates/src/main/kotlin/org/jetbrains/kotlin/gradle/benchmark/BenchmarkTemplate.kt#L133-L145